core/kernel/symbols.c

```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2019, EPAM Systems
+ */
+
+#include <stdio.h>
+#include <symbols.h>
+#include <util.h>
+
+/* All extern entries are auto-generated during first and second linking pass*/
+
+/* Symbol table */
+extern const struct syms_table syms_table[] __weak;
+/* Number of entries in syms_table */
+extern const unsigned int syms_count __weak;
+/* Compressed symbol names */
+extern const uint8_t syms_names[] __weak;
+/* Basically, absolute address of the first symbol */
+extern const vaddr_t syms_offset __weak;
+/* Indexes of symbols in the compressing table */
+extern const uint16_t syms_token_index[257] __weak;
+/* Compressing table */
+extern const char syms_token_table[] __weak;
+
+static const struct syms_table *find_symbol(vaddr_t addr, size_t *offset)
+{
+	unsigned int a, b;
+
+	if (addr < syms_offset)
+		return NULL;
+
+	addr -= syms_offset;
+
+	/*
+	 * Table is sorted and we assume that last entry points to end
+	 * of OP-TEE memory.
+	 */
+	if (addr > syms_table[syms_count - 1].offset)
+		return NULL;
+
+	/* Perform binary search */
+	a = 0;
+	b = syms_count;
+	while (b - a > 1) {
+		unsigned int idx = a + (b - a) / 2;
+
+		if (syms_table[idx].offset <= addr)
+			a = idx;
+		else
+			b = idx;
+	}
+
+	if (offset)
+		*offset = addr - syms_table[a].offset;
+
+	return syms_table + a;
+}
+
+static int syms_uncompress_name(const struct syms_table *sym, char *buf,
+				size_t buf_len)
+{
+	const uint8_t *data = syms_names + sym->name_offset;
+	size_t len = MIN(sym->name_len, buf_len);
+	int ret;
+
+	ret = len;
+	if (buf_len > len)
+		buf[len] = '\0';
+
+	while (len) {
+		const char *tok = syms_token_table + syms_token_index[*data];
+		size_t tok_len = syms_token_index[*data + 1] -
+			syms_token_index[*data];
+
+		for (; tok_len && len; tok_len--, len--)
+			*buf++ = *tok++;
+
+		data++;
+	}
+
+	return ret;
+}
+
+int syms_format_name_w_offset(vaddr_t addr, char *buf, size_t buf_len)
+{
+	size_t offset;
```

Comment:
I'm sorry, I don't get it. Could you please elaborate?

Reply:
sorry. I meant initialize each local variable where defined. See https://optee.readthedocs.io/en/latest/general/coding_standards.html#coding-standards, section that starts with `Variables are initialized according to these general guidelines:`.

Here it would make:
```
	size_t offset = 0;
	int len = 0;
	const struct syms_table *sym = find_symbol(addr, &offset);

	(...)

Reply:
Ah, I see. Looks like I somehow missed this addition to the coding style. Thank you for pointing out.